### ISSUE TITLE
WPF3D improvements.

### ISSUE THREAD

**Issue by darcome** [dotnet/wpf#2063]
Good Morning,

I'd like to know if there are any improvements planned for WPF3D to make it more similar to a modern 3D engine.

Thanks in advance

**Comment by lindexi** [comment]
https://github.com/dotnet/wpf/issues/2062

**Comment by weltkante** [comment]
@lindexi I think he's talking about the [Visual3D](https://docs.microsoft.com/de-de/dotnet/api/system.windows.media.media3d.visual3d?view=netframework-4.8) class subtree, D3DImage is only usefol for integrating external 3D engines, not for the builtin one.

@darcome I don't think anything is planned

**Comment by darcome** [comment]
Yes, I am referring about the internal one. @weltkante  thank you for your answer!

**Comment by grubioe** [comment]
Hi - we'll consider specific requests and proposals, perhaps you can expand on your ask? Thanks

**Comment by darcome** [comment]
Hi, I was thinking about the following, per example:
1) Make the engine use the latest version of DirectX instead of DX9.
2) Handle transparency automatically instead of having a user think what's the correct order to load objects in order for them to be transparent.
3) Give the possibility to write Vertex, Pixel and Geometry shaders.

These are just examples (I am not a 3D engine developer therefore I couldn't help).

Another possibility would be the following, even if probably is not feasable...

Would it be possible to sort of "port" BabylonJS to WPF so that the 3D Engine would be something modern as BabylonJS, but the simplicity to describe a 3D scene with XAML would remain? I know the BabylonJS team is a microsoft team, therefore maybe it would be possible to have the WPF3D team and the BabylonJS team cooperate?

I don't know... just my 2 cents...

**Comment by weltkante** [comment]
@darcome the engine already has been updated to use a newer version of DirectX, there was a blog post at one point about that, but I don't have a link ready and don't remember the exact Desktop Framework version with which the update happened.

**Comment by vatsan-madhavan** [comment]
WPF still uses DirectX 9. 

In .NET 4.7 I made a switch from [D3DX](https://en.m.wikipedia.org/wiki/D3DX) to DirectXMath. This brought in a dependency upon d3dcompiler_47.dll - which is a DX11 component (it supplies `D3DCompike` Win32 function IIRC). 

Besides this, WPF still uses DX9 in both .NET Framework and .NET core.

**Comment by vatsan-madhavan** [comment]
@darcome these are very good suggestions. I would suggest opening issues for each and let the community organically vote and help us prioritize each of these issues. 

> Give the possibility to write Vertex, Pixel and Geometry shaders.

It’a possible to incorporate your own shaders into a WPF application today. Have you had a chance to look into [ShaderEffect](https://docs.microsoft.com/en-us/dotnet/api/system.windows.media.effects.shadereffect) yet?

**Comment by weltkante** [comment]
> WPF still uses DirectX 9

Huh, now I'm very confused, I'm 100% sure there was a blog post about updating the WPF rendering engine. I even had posted a question whether D3DImage interop would be updated as well (answer was no but they'd forward the request). Did that effort exist and got canceled? I don't remember if there ever was an announcment that it was done, I may just have assumed it was because D3D11 started to get loaded into the process after some framework update.

That said I'll definitely try updating it once it finally releases, I've done something similar once already and it wasn't too hard. I just hope the closed source part (line services etc.) doesn't depend on DX9, that would really be a showstopper for me.

> It’a possible to incorporate your own shaders into a WPF application today. Have you had a chance to look into ShaderEffect yet?

I don't think that works on the 3D side of WPF, does it?

**Comment by mo-russo** [comment]
Hi, @weltkante @vatsan-madhavan Any news about the update to the rendering engine to support DirectX 11-12?

We currently have been trying to make Srgb work with D3DImage by looking into a few solutions like SlimDX, SharpDX, D3DImageEx but we believe that DirectX 9 does not support it. So now we need to copy the DirectX 11 surface to a memory buffer, pass it to WPF and use a WriteableBitmap to draw it. Which as you might think, has some performance issues especially on 4k monitors when our application runs fullscreen.

So we're wondering if we should hope for an official WPF update to support this or if we should investigate other solutions.

**Comment by weltkante** [comment]
I had a prototype reimplementing the WPF native side in D3D11 for rendering and D2D for geometry. It maps pretty cleanly, but its a reimplementation and not a port, so I can't judge the compatibility issues this would cause (lots of edge cases might behave differently). It was a crude prototype though, just proof of concept to check if its doable, and many of the fine details are not finished (in particular text). Having had a look at the source code I think porting it is not reasonable, but reimplementing the native library on D3D11/D2D and keeping the managed side should be reasonably doable.

I'm currently not proceeding with any of my WPF plans, since its obvious the team/repo is moving too slow for what I'm doing, so most likely my changes won't be flowing back, and I'm not ready for a fork since that would mean I need to have my own designer as well.

**Comment by lostmsu** [comment]
Do you know what's the status of the WPF native 3D backend? Is it even open source?

**Comment by lindexi** [comment]
@lostmsu Do you mean D3D ?

**Comment by lostmsu** [comment]
@lindexi nvm, I think I meant wpfgfx

**Comment by wave9d** [comment]
Any plans on updating WPF to DX11 ?

**Comment by lindexi** [comment]
@wave9d It doesn't seem necessary... At least for me

**Comment by uzername** [comment]
WPF 3D does not work well on Intel built-in videochips because Intel dropped support of DirectX 9 in their drivers.

It's really necessary

**Comment by lindexi** [comment]
@uzername Will it running on D3D9On12 ?

**Comment by uzername** [comment]
Feels like workaround for my cause, but I have no idea how to use it. I checked github page of D3D9On12 and I still have no idea how to use it in my C# project that shows 3D scene in WPF (to be more specific - I use HelixToolkit, but it is built on top of WPF 3D and it gets pretty close in here)

**Comment by uzername** [comment]
> @uzername Will it running on D3D9On12 ?

@lindexi 

I have checked this article: https://www.tomshardware.com/news/intel-xe-arc-swap-to-dx9-emulation
It states the following: 
> Native DX9 hardware support is [officially gone](https://www.intel.com/content/www/us/en/support/articles/000091238/graphics.html?s=31) from Intel's Xe integrated graphics solutions on 12th Gen CPUs and [A-Series Arc Alchemist](https://www.tomshardware.com/news/intel-arc-alchemist-release-date-specs-pricing-all-we-know) discrete GPUs. To replace it, all DirectX 9 support will be transferred to DirectX 12 in the form of emulation. Emulation will run on an open-source conversion layer known as "[D3D9On12](https://github.com/microsoft/D3D9On12)" from Microsoft.

That's why - I assume that WPF 3D uses D3D9On12 when launched on modern built-in Intel graphics chips. And it does not look okay. 
That means - migration to DX11 seems really necessary, at least for me  
. . . please . . .

**Comment by lindexi** [comment]
> That means - migration to DX11 seems really necessary, at least for me

I'm guessing you don't know what this means?

See https://github.com/microsoft/D3D11On12 LoL
